### PR TITLE
SPLICE-1222 Implement in-memory subtransactions (master)

### PR DESCRIPTION
**Pull request by dgomezferro**
Ported to master.

### CODE REVIEW

File: splice_si_api/src/main/java/com/splicemachine/si/impl/txn/WritableTxn.java [hunk 1]
```diff
@@ -129,46 +131,66 @@ public State getState(){
...
-                    throw exceptionFactory.cannotCommit(txnId,state);
+                    throw exceptionFactory.cannotCommit(txnId, state);
+            }
+            synchronized (this) {
+                //double-checked locking for efficiency--usually not needed
+                switch (state) {
+                    case COMMITTED:
+                        return;
+                    case ROLLEDBACK:
+                        throw exceptionFactory.cannotCommit(txnId, state);
```
Comment (snowspeeder) [inline splice_si_api/src/main/java/com/splicemachine/si/impl/txn/WritableTxn.java]: Nice to add default to switch stmt or convert to if-else if there are only 2 states

File: splice_si_api/src/main/java/com/splicemachine/si/impl/TransactionImpl.java [hunk 2]
```diff
@@ -60,58 +60,20 @@ public int setSavePoint(String name,Object kindOfSavepoint) throws IOException{
...
+//        if(LOG.isDebugEnabled())
+//            SpliceLogUtils.debug(LOG,"Before setSavePoint: name=%s, savePointStack=\n%s",name,getSavePointStackString());
         setActiveState(false,false,null); //make sure that we are active
         Txn currentTxn=getTxn();
         Txn child=lifecycleManager.beginChildTransaction(currentTxn,null);
         txnStack.push(Pair.newPair(name,child));
-        if(LOG.isDebugEnabled())
-            SpliceLogUtils.debug(LOG,"After setSavePoint: name=%s, savePointStack=\n%s",name,getSavePointStackString());
+//        if(LOG.isDebugEnabled())
+//            SpliceLogUtils.debug(LOG,"After setSavePoint: name=%s, savePointStack=\n%s",name,getSavePointStackString());
```
Comment (snowspeeder) [inline splice_si_api/src/main/java/com/splicemachine/si/impl/TransactionImpl.java]: should these debug messages be uncommented (also on #63,64) or removed?

File: splice_si_api/src/main/java/com/splicemachine/si/impl/txn/WritableTxn.java (hunk 1 above)
Comment (dgomezferro) [inline splice_si_api/src/main/java/com/splicemachine/si/impl/txn/WritableTxn.java]: Done

File: splice_si_api/src/main/java/com/splicemachine/si/impl/TransactionImpl.java (hunk 2 above)
Comment (dgomezferro) [inline splice_si_api/src/main/java/com/splicemachine/si/impl/TransactionImpl.java]: I removed them, now we end up with huge stacks that take a long time to serialize into a String.

**Pull request merged**